File: java/solitaire/klondike/src/main/java/org/silnith/game/solitaire/FoundationValidator.java
package org.silnith.game.solitaire;

import static org.silnith.deck.Value.ACE;

import java.util.Iterator;
import java.util.List;
import java.util.Map;

import org.silnith.deck.Card;
import org.silnith.deck.Suit;
import org.silnith.game.Validator;


/**
 * Validates that the foundation conforms to the rules of Klondike solitaire.
 * 
 * <p>The foundation is one stack for each {@link Suit}.  The stack for each suit
 * must start with the {@link org.silnith.deck.Value#ACE} and count up incrementally.</p>
 */
public class FoundationValidator implements Validator<Map<Suit, List<Card>>> {
    
    /**
     * Creates a new foundation validator.
     */
    public FoundationValidator() {
        super();
    }
    
    @Override
    public void validate(final Map<Suit, List<Card>> foundation) {
        for (final Suit suit : Suit.values()) {
            final List<Card> cardsForSuit = foundation.get(suit);
            if (cardsForSuit == null) {
                throw new IllegalArgumentException("No list of cards for suit " + suit);
            }
            final Iterator<Card> iterator = cardsForSuit.iterator();
            if (iterator.hasNext()) {
                Card previousCard = iterator.next();
                validateSuit(suit, previousCard);
                
                if (previousCard.getValue() != ACE) {
                    throw new IllegalArgumentException("First card in foundation for suit " + suit + " must be an " + ACE + ", not a " + previousCard + ".");
                }
                while (iterator.hasNext()) {
                    final Card currentCard = iterator.next();
                    validateSuit(suit, currentCard);
                    
                    if (previousCard.getValue().getValue() + 1 != currentCard.getValue().getValue()) {
                        throw new IllegalArgumentException("Cannot put " + currentCard + " on top of " + previousCard + " in foundation for suit " + suit + ".");
                    }
                    previousCard = currentCard;
                }
            }
        }
    }
    
    private void validateSuit(final Suit suit, final Card card) {
        if (card.getSuit() != suit) {
            throw new IllegalArgumentException("Cannot have " + card + " in foundation for suit " + suit + ".");
        }
    }
    
}

File: csharp/Solitaire/Klondike/FoundationValidator.cs
using Silnith.Game.Deck;
using System;
using System.Collections.Generic;

namespace Silnith.Game.Klondike
{
    /// <summary>
    /// Validates that the foundation conforms to the rules of Klondike solitaire.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The foundation is one stack for each <see cref="Suit"/>.  The stack for each suit
    /// must start with the <see cref="Value.Ace"/> and count up incrementally.
    /// </para>
    /// </remarks>
    public class FoundationValidator : IValidator<IReadOnlyDictionary<Suit, IReadOnlyList<Card>>>
    {
        private readonly IEnumerable<Suit> AllSuits = new List<Suit>()
        {
            Suit.Club,
            Suit.Diamond,
            Suit.Heart,
            Suit.Spade,
        };

        /// <inheritdoc/>
        public void Validate(IReadOnlyDictionary<Suit, IReadOnlyList<Card>> foundation)
        {
            foreach (Suit suit in AllSuits)
            {
                if (foundation.TryGetValue(suit, out IReadOnlyList<Card> cardsForSuit))
                {
                    IEnumerator<Card> enumerator = cardsForSuit.GetEnumerator();
                    if (enumerator.MoveNext())
                    {
                        // there is something, check it
                        Card previousCard = enumerator.Current;
                        ValidateSuit(suit, previousCard);
                        if (previousCard.Value != Value.Ace)
                        {
                            throw new ArgumentException("First card in foundation for suit " + suit + " must be " + Value.Ace + " instead it is " + previousCard + ".");
                        }
                        while (enumerator.MoveNext())
                        {
                            Card currentCard = enumerator.Current;
                            ValidateSuit(suit, currentCard);

                            if (previousCard.Value.GetValue() + 1 != currentCard.Value.GetValue())
                            {
                                throw new ArgumentException("Cannot put " + currentCard + " on top of " + previousCard + " in foundation for suit " + suit + ".");
                            }

                            previousCard = currentCard;
                        }
                    }
                }
                else
                {
                    throw new ArgumentException("No foundation entry for " + suit);
                }
            }
        }

        private void ValidateSuit(Suit suit, Card card)
        {
            if (card.Suit != suit)
            {
                throw new ArgumentException("Cannot have " + card + " in foundation for suit " + suit + ".");
            }
        }
    }
}
